grpc-java/grpc-servlet-websocket-jakarta/src/main/java/io/grpc/servlet/web/websocket/MultiplexedWebSocketServerStream.java

```diff
@@ -54,12 +67,55 @@ public class MultiplexedWebSocketServerStream extends AbstractWebSocketServerStr
     private final Map<Integer, MultiplexedWebsocketStreamImpl> streams = new HashMap<>();
     private final boolean isTextRequest = false;// not supported yet
 
+    /**
+     * Enum to describe the process of closing a transport. After the server has begun to close but the client hasn't
+     * yet acknowledged, it is permitted for the client to start a new stream, but after the server acknowledges, no new
+     * streams can be started. Note that shutdown will proceed anyway, and will eventually stop all streams.
+     */
+    enum ClosedState {
+        OPEN, CLOSING, CLOSED
+    }
+
+    private ClosedState closed = ClosedState.OPEN;
+    private CompletableFuture<Void> closingFuture = new CompletableFuture<>();
+
     public MultiplexedWebSocketServerStream(ServerTransportListener transportListener,
             List<? extends ServerStreamTracer.Factory> streamTracerFactories, int maxInboundMessageSize,
             Attributes attributes) {
         super(transportListener, streamTracerFactories, maxInboundMessageSize, attributes);
     }
 
+    /**
+     * Stops this multiplexed transport from accepting new streams. Instead, it will reply with its version of GO_AWAY,
+     * a stream of Integer.MAX_INTEGER to the client to signal that new requests will not be accepted, and future
+     * incoming streams will be closed by the server right away. In keeping with h2, until the client ACKs the close, we
+     * will permit incoming streams that are were sent before we closed, but they likely will not have a large window to
+     * get their work done before they are closed the rest of the way.
+     */
+    private CompletableFuture<Void> stopAcceptingNewStreams() {
+        if (closed != ClosedState.OPEN) {
+            return closingFuture;
+        }
+        closed = ClosedState.CLOSING;
```

Comment:
Should `closed` technically be marked as `volatile`? Even though we expect `closed` to go OPEN -> CLOSING -> CLOSED with happens-before/happens-after semantics, I don't think the java memory model guarantees other threads will see the change?

Reply:
Ah, I thought I had this set to only be possible from one thread, but I will double check and get back to you, or at least try to tell a convincing story...

Reply:
Yes, that could be helpful, but I think not strictly necessary - `stopAcceptingNewStreams()` can only be called once, and the other tests require a full round trip to the client after that to make sense. I'm not sure what the threading "guarantees" there would be, except that by the time we've sent a message and received a corresponding reply, that change should have propagated. 

The exception to this is a newly opened socket, but I think that onOpen is already racing WebsocketFactory's loop in that case. Will investigate.

Open to adding it anyway, but in that case we have many more fields that should h